### ISSUE TITLE
Constrained Recall Objective has weird interaction with LightGBM early stopping criteria

### ISSUE THREAD

**Issue by AndreFCruz** [feedzai/fairgbm#11]
## Description

- When optimizing for Recall (minimizing FNR), only label positive samples are considered for computing the loss or its gradient;
- However, passing a gradient of zero for all label negatives leads to weird behavior in the `GBM::Train` function;
- So, for now, we're scaling down the gradient of all label negatives by multiplying them with a tiny positive number: see the `label_negative_weight` in `ConstrainedRecallObjective::GetGradients`;
  - This shouldn't be needed, but seems to temporarily fix the issue with no unintended consequences (as the gradient flowing is _very_ small);

## Reproducible example

1. Omit the else clause in `ConstrainedRecallObjective::GetGradients`, which deals with label negative samples, and in theory should not be needed for optimizing for recall;
2. Compile and run, and observe weird "-inf split" messages, which can lead to training stopping too early;

**Comment by daniel-n-silva** [comment]
The current fix (`label_negative_weight = 1e-2` applied to `z * (1-z)`) resolves the `-inf split` this issue reports, but I think the safety margin it provides is not principled and erodes as training progresses — worth flagging even though it doesn't reproduce the exact original failure mode.

**Why the margin erodes**

The per-sample hessian for label negatives is `label_negative_weight * z * (1 - z)`, where `z = sigmoid(score + xent_horizontal_shift)`. Note the FPR constraint doesn't contribute to this hessian either — `GetConstraintGradientsWRTModelOutput` takes an unused `hessians` parameter ([constrained_objective_function.h:257-258](https://github.com/feedzai/fairgbm/blob/main-fairgbm/include/LightGBM/constrained_objective_function.h#L257-L258)), consistent with `gbdt.cpp:233` ("Dump hessians, we don't currently use them though :P") — so the full hessian budget for a label-negative sample really is just `1e-2 * z*(1-z)`.

That factor is only safely above `min_sum_hessian_in_leaf` (`1e-3` by default, [config.h:305](https://github.com/feedzai/fairgbm/blob/main-fairgbm/include/LightGBM/config.h#L305)) while `score` stays close to 0. As the model becomes confident about a negative — which is the normal, desired outcome for "obvious" non-fraud cases after enough boosting rounds — `z` saturates toward 0 and the hessian collapses along with it. With the current defaults (`stepwise_proxy_margin=1.0` → `xent_horizontal_shift≈0.5413`, `label_negative_weight=1e-2`, `min_data_in_leaf=20`):

```
score   z=sigmoid(score+shift)   hessian/sample   sum over 20 samples
  -4         0.0305                2.96e-04            5.92e-03
  -5         0.0114                1.13e-04            2.26e-03
  -6         0.0042                4.22e-05            8.45e-04   <- below 1e-3
  -7         0.0016                1.56e-05            3.12e-04   <- below 1e-3
  -8         0.0006                5.76e-06            1.15e-04   <- below 1e-3
```
(crossing point for a 20-sample leaf: `score ≈ -5.83`; a logodds of -6 to -8 is a predicted probability of roughly 0.03%-0.25%, an ordinary outcome for a well-separated negative)

**What this does and doesn't mean**

To be clear about scope: this is static code analysis plus the arithmetic above, not a training repro. And it's not a reappearance of the original bug — that happened when the *entire tree* found no valid split anywhere (`num_leaves() == 1`), which requires the whole dataset's hessian mass to collapse; with realistic dataset sizes that stays far above the threshold even at very negative scores. What I think does happen is narrower: individual leaves that end up pure (or nearly pure) label-negative, at scores past roughly -6, become insplittable — not a training-halting event, but a per-leaf ceiling on how far the tree can specialize in confidently-negative regions, and one whose reach silently depends on how confident the model has already become. That seems like exactly the kind of state-dependent behavior behind the "prone to high variability of outcomes" warning already logged at construction time ([constrained_recall_objective.hpp:70-71](https://github.com/feedzai/fairgbm/blob/main-fairgbm/src/objective/constrained_recall_objective.hpp#L70-L71)).

**On alternative fixes**

The obvious first idea — decoupling the hessian from `z*(1-z)` with a constant floor — isn't as neutral as it looks: with `lambda_l2=0`/`max_delta_step=0` (both defaults), the leaf output for a pure-LN leaf today is `-Σg/Σh ≈ -z̄/(z̄(1-z̄)) ≈ -1`; a constant hessian would change that to `≈ -z̄ ≈ 0`, which is a real change to the Newton step on label negatives, not just a safety guard. Any fix here needs to account for that rather than assume it's free.

Wanted to flag the reasoning in case it's useful, given the objective is already marked as under active research — happy to help investigate further if there's interest.
